### PR TITLE
feat: add support for header based route

### PR DESCRIPTION
**Pull request by shawnh2**
Add support for header based route.
- when use it alone in a route, it has the higher priority to be matched in a router

### CODE REVIEW

File: pixiu/pkg/common/router/router.go [hunk 1]
```diff
@@ -83,6 +114,21 @@ func (rm *RouterCoordinator) initTrie() {
...
 
+func (rm *RouterCoordinator) initRegex() {
+	for _, router := range rm.activeConfig.Routes {
+		headers := router.Match.Headers
+		for i := range headers {
+			if headers[i].Regex && len(headers[i].Values) > 0 {
+				// regexp always use first value of header
+				err := headers[i].SetValueRegex(headers[i].Values[0])
+				if err != nil {
+					logger.Errorf("invalid regexp in headers[%d]: %v", i, err)
```
Comment (mark4z) [inline pixiu/pkg/common/router/router.go:126]: Shouldn't it be panic here?

**Review by mark4z — APPROVED** [review]

**Comment by mark4z** [comment]
考虑扩展下现有的samples来演示如何使用header路由

**Comment by mark4z** [comment]
Good job!

File: pixiu/pkg/common/router/router.go (hunk 1 above)
Comment (shawnh2) [reply pixiu/pkg/common/router/router.go:126]: fixed

**Comment by shawnh2** [comment]
> 考虑扩展下现有的samples来演示如何使用header路由

sample has been added, https://github.com/apache/dubbo-go-pixiu-samples/pull/30

**Comment by AlexStocks** [comment]
@yqxu review this pr

File: pixiu/pkg/model/router.go [hunk 2]
```diff
@@ -75,22 +75,83 @@ func NewRouterMatchPrefix(name string) RouterMatch {
 	return RouterMatch{Prefix: "/" + name + "/"}
 }
 
-func (rc *RouteConfiguration) RouteByPathAndMethod(path, method string) (*RouteAction, error) {
+func (rc *RouteConfiguration) RouteByPathAndMethod(path, method string, header stdHttp.Header) (*RouteAction, error) {
```
Comment (yqxu) [inline pixiu/pkg/model/router.go]: The behavior of a method should be reflected in its name. It is recommended to modify the name as RouteByPathAndMethodAndHeader.

File: pixiu/pkg/common/router/trie/trie.go [hunk 3]
```diff
@@ -317,7 +339,7 @@ func (node *Node) putPathVariable(pathVariable string, isReal bool, bizInfo inte
 	return true
 }
 
-func (node *Node) putNode(matchStr string, isReal bool, bizInfo interface{}) bool {
+func (node *Node) putNode(matchStr string, isReal bool, bizInfo interface{}, header map[string][]string) bool {
```
Comment (yqxu) [inline pixiu/pkg/common/router/trie/trie.go]: The "header" should be an attribute of "bizInfo".

File: pixiu/pkg/common/router/trie/trie.go [hunk 4]
```diff
@@ -333,13 +355,14 @@ func (node *Node) putNode(matchStr string, isReal bool, bizInfo interface{}) boo
 
 	if isReal {
 		selfNode.bizInfo = bizInfo
+		selfNode.header = header
```
Comment (yqxu) [inline pixiu/pkg/common/router/trie/trie.go]: The "header" should be an attribute of "bizInfo".

**Review by yqxu — APPROVED** [review]

**Pull request merged**